xds/src/test/java/io/grpc/xds/internal/certprovider/MeshCaCertificateProviderTest.java

```diff
@@ -0,0 +1,538 @@
+   */
+  private static final long CERT0_EXPIRY_TIME_MILLIS = 1899853658000L;
+  /**
+   * Cert validity of 12 hours for the above cert.
+   */
+  private static final long CERT0_VALIDITY_MILLIS = TimeUnit.MILLISECONDS
+      .convert(12, TimeUnit.HOURS);
+  /**
+   * Compute current time based on cert expiry and cert validity.
+   */
+  private static final long CURRENT_TIME_NANOS =
+      TimeUnit.MILLISECONDS.toNanos(CERT0_EXPIRY_TIME_MILLIS - CERT0_VALIDITY_MILLIS);
+  @Rule
+  public final GrpcCleanupRule cleanupRule = new GrpcCleanupRule();
+
+  private static class ResponseToSend {
+    Throwable getThrowable() {
+      throw new UnsupportedOperationException("Called on " + getClass().getCanonicalName());
+    }
+
+    List<String> getList() {
+      throw new UnsupportedOperationException("Called on " + getClass().getCanonicalName());
+    }
+  }
+
+  private static class ResponseThrowable extends ResponseToSend {
+    final Throwable throwableToSend;
+
+    ResponseThrowable(Throwable throwable) {
+      throwableToSend = throwable;
+    }
+
+    @Override
+    Throwable getThrowable() {
+      return throwableToSend;
+    }
+  }
+
+  private static class ResponseList extends ResponseToSend {
+    final List<String> listToSend;
+
+    ResponseList(List<String> list) {
+      listToSend = list;
+    }
+
+    @Override
+    List<String> getList() {
+      return listToSend;
+    }
+  }
+
+  private final Queue<Meshca.MeshCertificateRequest> receivedRequests = new ArrayDeque<>();
+  private final Queue<String> receivedStsCreds = new ArrayDeque<>();
+  private final Queue<String> receivedZoneValues = new ArrayDeque<>();
+  private final Queue<ResponseToSend> responsesToSend = new ArrayDeque<>();
+  private final Queue<String> oauth2Tokens = new ArrayDeque<>();
+  private final AtomicBoolean callEnded = new AtomicBoolean(true);
+
+  @Mock private MeshCertificateServiceGrpc.MeshCertificateServiceImplBase mockedMeshCaService;
+  @Mock private CertificateProvider.Watcher mockWatcher;
+  @Mock private BackoffPolicy.Provider backoffPolicyProvider;
+  @Mock private BackoffPolicy backoffPolicy;
+  @Spy private GoogleCredentials oauth2Creds;
+  @Mock private ScheduledExecutorService timeService;
+  @Mock private TimeProvider timeProvider;
+
+  private ManagedChannel channel;
+  private MeshCaCertificateProvider provider;
+
+  @Before
+  public void setUp() throws IOException {
+    MockitoAnnotations.initMocks(this);
+    when(backoffPolicyProvider.get()).thenReturn(backoffPolicy);
+    when(backoffPolicy.nextBackoffNanos())
+        .thenReturn(DELAY_VALUES[0], DELAY_VALUES[1], DELAY_VALUES[2]);
+    doAnswer(
+        new Answer<AccessToken>() {
+          @Override
+          public AccessToken answer(InvocationOnMock invocation) throws Throwable {
+            return new AccessToken(
+                oauth2Tokens.poll(), new Date(System.currentTimeMillis() + 1000L));
+          }
+        })
+        .when(oauth2Creds)
```

Comment:
Why not just implement an extension of `GoogleCredentials`(it has a public and a protected constructor) and override `refreshAccessToken()`

Reply:
Ack. With a Mock/Spy I can get rid of a lot of boilerplate code (create a subclass of GoogleCredentials etc) and use a concise notation to inject test data which is why I prefer it.

Reply:
There is not much boilerplate for GoogleCredentials. It will be a lot cleaner than `doAnswer`.

Reply:
Okay I'll consider refactoring and using a test class for this in my next PR